NebulaNetwork/Ngrok/NgrokManager.cs

```diff
@@ -0,0 +1,224 @@
+                    if (!hasDownloadAndInstallBeenConfirmed)
+                    {
+                        NebulaModel.Logger.Log.Warn("Failed to download or install Ngrok, because user rejected Ngrok download and install confirmation!");
+                        return;
+                    }             
+                    
+                    try
+                    {
+                        await DownloadAndInstallNgrok();
+                    }
+                    catch
+                    {
+                        NebulaModel.Logger.Log.Warn("Failed to download or install Ngrok!");
+                        throw;
+                    }
+
+                }
+
+                if (!StartNgrok())
+                {
+                    NebulaModel.Logger.Log.Warn("Failed to start Ngrok tunnel!");
+                    return;
+                }
+
+                if (!IsNgrokStarted())
+                {
+                    NebulaModel.Logger.Log.Warn("Ngrok tunnel has exitted prematurely! Invalid authtoken perhaps?");
+                    return;
+                }
+
+                try
+                {
+                    NgrokAddress = await GetTunnelAdress();
+                }
+                catch
+                {
+                    NebulaModel.Logger.Log.Warn("Failed to obtain Ngrok address!");
+                    throw;
+                }
+
+            });
+
+        }
+
+        public async Task DownloadAndInstallNgrok()
+        {
+            using (var client = new HttpClient())
+            {
+                using (var s = client.GetStreamAsync("https://bin.equinox.io/c/bNyj1mQVY4c/ngrok-v3-stable-windows-amd64.zip"))
+                {
+                    using (var zip = new ZipArchive(await s, ZipArchiveMode.Read))
+                    {
+                        zip.ExtractToDirectory(Path.GetDirectoryName(_ngrokPath));
+                    }
+                }
+            }
+        }
+
+        public bool IsNgrokInstalled()
+        {
+            return File.Exists(_ngrokPath);
+        }
+
+        public bool StartNgrok()
+        {
+            StopNgrok();
+
+            _ngrokProcess = new Process();
+            _ngrokProcess.StartInfo = new ProcessStartInfo
+            {
+                WindowStyle = ProcessWindowStyle.Hidden,
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                UseShellExecute = false,
+                CreateNoWindow = true,
+                FileName = _ngrokPath,
+                Arguments = $"tcp {_port} --authtoken {_authtoken}" + (!string.IsNullOrEmpty(_region) ? $" --region {_region}" : ""),
+            };
+            return _ngrokProcess.Start();
+        }
+
+        public void StopNgrok()
+        {
+            if (_ngrokProcess != null)
+            {
+                _ngrokProcess.Refresh();
+                if (!_ngrokProcess.HasExited)
+                {
+                    _ngrokProcess.Kill();
```

Comment:
```suggestion
                    _ngrokProcess.CloseMainWindow();
```

Should we use CloseMainWindow instead of Kill to gracefully shutdown the process?

Reply:
To be honest, I'm more comfortable with just killing the process (this is also how the other API package does it), since CloseMainWindow is more voluntary and thus might be blocked by ngrok for some reason (leaving a dangling process or blocking execution).